README.md
# python-quick-network-tool
python-quick-network-tool

Author: DJENGINEER

For when you need quick **HTTP** and **FTP** servers for file transfers. HTTP and FTP data transfers are not encrypted. Do not use if you require encryption.

HTTPS and SFTP is available from V1.2.2. Do not use on production servers.

The IP address in the second section may differ due to different number of network adapters or adapter names. If you are using for local network, then use your local network ip address, which may start with 192.168.\*\*\*.\*\*\*

### Regarding HTTPS and SFTP
If you run HTTPS server instead of HTTP, specify **https://** when accessing the url like this **https://[ip_address]:8000**.

If you run TLS/SSL enabled HTTP and FTP, your browser and FTP client may show a certificate warning. You may accept the warning and proceed with the connection.

![python quick network tool](https://github.com/djengineer/python-quick-network-tool/blob/main/screenshot%20v1.2.3.jpg?raw=true)



# Binaries
The tools are in the binaries folder. The windows binary may be flagged by anti-viruses as Trojan (AVG will do this). From V1.2.2, the binaries are properly packed with UPX. Windows may give a warning when running an unknown .exe file. 

## v1.2.4 Binaries MD5 Checksum

py-quick-network-tool-ubuntu-64-v1-2-4.bin: 4c189899b24bb86ec5b4d4af1e9ca856
py-quick-network-tool-win-64-v1-2-4.exe: f2fc04441ca82bd06ad55a51e2f94a8c

## v1.2.3 Binaries MD5 Checksum

py-quick-network-tool-ubuntu-64-v1-2-3.bin: 1deef0e2afb77f27a9f782df7e20e287
py-quick-network-tool-win-64-v1-2-3.exe: 4c3dce6e998d4d4c142989b142f12a06

## v1.2.2 Binaries MD5 Checksum
py-quick-network-tool-ubuntu-64-v-1-2-2.bin: 92d5c2d69462696fc6ce5738d858528e

py-quick-network-tool-win-64-v-1-2-2.exe: 968ee1b952db4664c55479d3357f6150


# Running as script Windows or Linux

Uncomment "build_for" when building for windows or linux.
The difference is windows will use the https://pypi.org/project/hdpitkinter/ for proper scaling to avoid blur text. PSUTIL list sequence is also different when listing network interfaces

```python
# create Python3 virtual environment first.
pip install -r requirements.txt
python app.py

```

# Building from source
```python
pip install -r requirements.txt
pyinstaller -F --noconsole app.py -n py-quick-network-tool
pyinstaller -F --noconsole app.py --upx-dir="./upx-3.96-win64/" -n py-quick-network-tool
pyinstaller -F --noconsole app.py --upx-dir="./upx-3.96-amd64_linux/" -n py-quick-network-tool

pyinstaller py-quick-network-tool.spec --upx-dir="./upx-3.96-win64/"
pyinstaller py-quick-network-tool.spec --upx-dir="./upx-3.96-amd64_linux/"

```

# Certificates
## HTTPS
To generate your own certificate for HTTPS, use the following line.
```bash
# no passphrase
openssl req -nodes -new -x509  -keyout key.pem -out cert.pem
```
## Certificate for pyftpd 
https://pyftpdlib.readthedocs.io/en/latest/tutorial.html#ftps-ftp-over-tls-ssl-server

We use the demo certificate referenced in the documentation. https://github.com/giampaolo/pyftpdlib/blob/master/demo/keycert.pem

If you want to create a new certificate, use the following commands.

```bash
# combining both private key and certificate into one pem file
openssl req -nodes -new -x509  -keyout key.pem -out cert.pem
cat cert.pem key.pem > server.includesprivatekey.pem
# change the name to the appropriate ftp cert name in the codes
```

# DEV STATUS

Python script and binaries app.py works well in both windows and linux.






changelog.md
# Change Log

v1.2.4
- Changed HTTP server from Simple to Threaded to support concurrent requests.
- Included PyTest for testing. Does not include testing for concurrent request.


v1.2.3
- Rearranged HTTP and FTP buttons
- Fixed pyinstaller certificate pathing. HTTPS and SFTP fixed
- Error handling if the server cannot be started because of duplicate processes.

v1.2.2
- FTP server is now enabled with TLS. Certificate used is from https://github.com/giampaolo/pyftpdlib/blob/master/demo/keycert.pem given by the official pyftpd documentation https://pyftpdlib.readthedocs.io/en/latest/tutorial.html#ftps-ftp-over-tls-ssl-server
- HTTPS server option is now available.
- PyOpenssl 22.0.0 is installed with pip for SFTP. 22.1.0(current latest) will result in an error.
- HTTP and FTP server status color coded.
- Changed the pyinstaller command. Previous binaries were not packed due to wrong command in README.

v1.2.1
- script will detect if operating is windows or linux. Do not have to manually specify the os when creating the binary with pyinstaller.
- Tkinter window height changed to 500px so that windows can display all items without scrolling.

tests.py
import pytest
import time
import os
import requests
from app import start_http_button,start_https_button,stop_http_button,start_ftp_button,start_ftps_button,stop_ftp_button
import ftplib
from ftplib import FTP,FTP_TLS

# create 100mb file for test
size = 100000000
with open("./BigTestFile.txt", "wb") as f:
    f.write(b" " * size)
with open("./SmallTestFile.txt", "wb") as f:
    f.write(b" " * 1000)
print("Temp files created for testing.")


url = 'http://127.0.0.1:8000'
https_url = 'https://127.0.0.1:8000'

class TestClassHTTP:
	"""
	HTTP status code tests
	"""
	def test_http(self):
		start_http_button()
		x = requests.get(url)
		stop_http_button()
		assert x.status_code == 200
	def test_https(self):
		start_https_button()
		x = requests.get(https_url, verify=False)
		stop_http_button()
		assert x.status_code == 200
	def test_small_file(self):
		start_http_button()
		x = requests.get(url+"/SmallTestFile.txt")
		stop_http_button()
		assert x.status_code == 200
	def test_big_file(self):
		start_http_button()
		x = requests.get(url+"/BigTestFile.txt")
		stop_http_button()
		assert x.status_code == 200

class TestClassFTP:
	def test_login(self):
		start_ftp_button()
		time.sleep(3)
		ftp = FTP()  # connect to host, default port
		ftp.connect('127.0.0.1',8021)
		result = ftp.login("user123","pass123")
		stop_ftp_button()
		assert result == "230 Login successful."
	def test_ftps_login(self):
		start_ftps_button()
		time.sleep(3)
		ftps = FTP_TLS()  # connect to host, default port
		ftps.connect('127.0.0.1',8021)
		result = ftps.login("user123","pass123")
		stop_ftp_button()
		assert result == "230 Login successful."
class TestClassTeardown:
	def test_teardown(self):		
		print("REMOVING FILES AND STOPPING SERVERS")
		os.remove("./BigTestFile.txt")
		os.remove("./SmallTestFile.txt")
		stop_http_button()
		stop_ftp_button()

app backup.py
from tkinter import *

import threading
import http.server
import socket 
from http.server import HTTPServer, SimpleHTTPRequestHandler, BaseHTTPRequestHandler
import http.server
import http.server
import os

import multiprocessing

from pyftpdlib.authorizers import DummyAuthorizer
from pyftpdlib.handlers import FTPHandler
from pyftpdlib.servers import FTPServer
from pyftpdlib.servers import ThreadedFTPServer

import threading
web_dir = os.path.join(os.path.dirname(__file__), './')
os.chdir(web_dir)
class ThreadedHttpd (threading.Thread):
	def __init__(self, threadID, name, ):
		threading.Thread.__init__(self)
		self.threadID = threadID
		self.name = name
		self.start_status = False
	def run(self,server_class=HTTPServer, handler_class=SimpleHTTPRequestHandler):
		server_address = ('', 8000)
		httpd = server_class(server_address, handler_class)
		httpd.serve_forever()
		
	def start_server(self):
		self.server_process = multiprocessing.Process(target=self.run())
		self.server_process.start()
	def shutdown(self):
		self.server_process.terminate()



def ftp_server_start():
    authorizer = DummyAuthorizer()
    authorizer.add_user('user', '12345', '.')
    handler = FTPHandler
    handler.authorizer = authorizer
    server = ThreadedFTPServer(('', 2121), handler)
    server.serve_forever()



http_server = ThreadedHttpd(1, "Thread-1")

window = Tk()
label = Label(window, text="I am a label widget")
start_http = Button(window, text="start Http port 8080",command=lambda:http_server.start_server())
stop_http = Button(window, text="stop http server",command=lambda:http_server.shutdown())


#start_ftp = Button(window, text="start FTP server",command=ftp_server_start)
#stop_ftp = Button(window, text="stop fto server",command=ftp_server_shutdown)


label.pack()
start_http.pack()
stop_http.pack()

if __name__ == "__main__":
	window.mainloop()

app.py
from tkinter import *

import threading
import http.server
import socket 
from http.server import HTTPServer, SimpleHTTPRequestHandler, BaseHTTPRequestHandler, ThreadingHTTPServer
import http.server

import multiprocessing

from pyftpdlib.authorizers import DummyAuthorizer
from pyftpdlib.handlers import FTPHandler
from pyftpdlib.handlers import TLS_FTPHandler
from pyftpdlib.servers import FTPServer
from pyftpdlib.servers import ThreadedFTPServer


from tkinter import messagebox
from tkinter import *

from hdpitkinter import HdpiTk
import psutil

import threading
import ssl

#web_dir = os.path.join(os.path.dirname(__file__), './')
#os.chdir(web_dir)

import sys, os

version_number = "1.2.4"
################################
##### Windows or Linux #########
################################
# change uncomment the appropriate build_for when building
#build_for ="linux"

build_for = os.name
if build_for == "posix":
	window = Tk()
elif build_for == "nt":
	window = HdpiTk()

global application_path

if getattr(sys, 'frozen', False):
	# If the application is run as a bundle, the PyInstaller bootloader
	# extends the sys module by a flag frozen=True and sets the app 
	# path into variable _MEIPASS'.
	#application_path = sys._MEIPASS
	application_path = os.path.dirname(sys.executable)
else:
	application_path = os.path.dirname(os.path.abspath(__file__))


global certdir
if hasattr(sys, "_MEIPASS"):
	# this is for pyinstaller's bundled file
    base_dir = os.path.abspath(os.path.dirname(__file__))
else:
	# if this is not for bundled file, just for the script
    base_dir = './'

global http_server_process
global ftp_server_process
global http_status
global ftp_status
http_server_process = None
ftp_server_process = None
http_status = "STOPPED"
ftp_status = "STOPPED"

class MyHttpHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=application_path, **kwargs)


# HTTPServer for non-threaded. ThreadingHTTPServer for threaded.
def httpd(server_class=ThreadingHTTPServer, handler_class=MyHttpHandler):
	server_address = ('', 8000)
	httpd = server_class(server_address, handler_class)
	httpd.serve_forever()

def httpsd(server_class=ThreadingHTTPServer, handler_class=MyHttpHandler):
	global certdir
	server_address = ('', 8000)
	httpd = server_class(server_address, handler_class)
	httpd.socket = ssl.wrap_socket (httpd.socket, 
        keyfile=base_dir + "/certificates/key.pem", 
        certfile=base_dir + '/certificates/cert.pem', server_side=True)
	httpd.serve_forever()

def start_http_button():
	global http_server_process
	global http_status
	if http_server_process != None:
		print("HTTP/HTTPS server already started.")
		pass
	elif http_server_process == None:
		# Name error means not defined. start a new server.
		http_server_process = multiprocessing.Process(target=httpd,name='HTTPprocess')
		http_server_process.start()
		http_status = "STARTED"
		print("HTTP server process started")

def start_https_button():
	global http_server_process
	global http_status
	if http_server_process != None:
		print("HTTP/HTTPS server already started.")
		pass
	elif http_server_process == None:
		# Name error means not defined. start a new server.
		http_server_process = multiprocessing.Process(target=httpsd,name='HTTPSprocess')
		http_server_process.start()
		http_status = "STARTED"
		print("HTTPS server process started")

def stop_http_button():
	global http_server_process
	global http_status
	if http_server_process == None:
		print("No server to stop.")
		pass
	else:
		http_server_process.terminate()
		http_server_process = None
		http_status = "STOPPED"
		print("HTTP server process terminated")

def ftpd():
    authorizer = DummyAuthorizer()
    authorizer.add_user('user123', 'pass123', '.', perm='elradfmwMT')
    handler = FTPHandler
    #handler = TLS_FTPHandler
    #handler.certfile = 'keycert.pem'
    handler.authorizer = authorizer
    server = ThreadedFTPServer(('0.0.0.0', 8021), handler)
    server.serve_forever()
def ftpsd():
    authorizer = DummyAuthorizer()
    authorizer.add_user('user123', 'pass123', '.', perm='elradfmwMT')
    handler = TLS_FTPHandler
    handler.certfile = base_dir + '/certificates/keycert_ftp.pem'
    handler.authorizer = authorizer
    server = ThreadedFTPServer(('0.0.0.0', 8021), handler)
    server.serve_forever()

def start_ftp_button():
	global ftp_server_process
	if ftp_server_process != None:
			print("FTP server already started.")
			pass
	elif ftp_server_process == None:
		# Name error means not defined. start a new server.
		ftp_server_process = multiprocessing.Process(target=ftpd,name='FTPprocess')
		ftp_server_process.start()
		print("FTP Server process started")
		global ftp_status
		ftp_status = "STARTED"

def start_ftps_button():
	global ftp_server_process
	if ftp_server_process != None:
			print("FTP server already started.")
			pass
	elif ftp_server_process == None:
		# Name error means not defined. start a new server.
		ftp_server_process = multiprocessing.Process(target=ftpsd,name='FTPprocess')
		ftp_server_process.start()
		print("FTPS Server process started")
		global ftp_status
		ftp_status = "STARTED"

def stop_ftp_button():
	global ftp_server_process
	global ftp_status
	if ftp_server_process == None:
		print("No server to stop.")
		pass
	else:
		ftp_server_process.terminate()
		ftp_server_process = None
		ftp_status = "STOPPED"
		print("FTP server process terminated")


def on_closing():
	global http_server_process
	global ftp_server_process
	if messagebox.askokcancel("Quit", "Do you want to close all services and quit?"):
		try:
			http_server_process.terminate()
		except NameError:
			# if no server process, skip. NameError if server process not started
			pass
		except:
			pass

		try:
			ftp_server_process.terminate()
		except NameError:
			# if no server process, skip. NameError if server process not started
			pass
		except:
			pass
		window.destroy()

network_info = psutil.net_if_addrs()
#print(network_info)
network_list = []
interface_details = ""
if build_for == "posix":
	for interface in network_info:
		temp_string = interface+": "+network_info[interface][0][1] +"\n"
		interface_details = interface_details + temp_string
		network_list.append([interface,network_info[interface][0][1]])
elif build_for == "nt":
	for interface in network_info:
		temp_string = interface+": "+network_info[interface][1][1] +"\n"
		interface_details = interface_details + temp_string
		network_list.append([interface,network_info[interface][1][1]])


if build_for == "posix":
	IPAddr1=network_list[1][1]
	#network_label = Entry(window, text=interface_details,  justify=RIGHT)
	instruction_text = "\nAccessing HTTP server in browser: http://%s:8000\nAccessing FTP server:\nUsername: user123\nPassword: pass123\nPort: 8021\nCommand in Linux: ftp %s 8021" % (IPAddr1,IPAddr1)

elif build_for == "nt":
	IPAddr1=network_list[0][1]
	#network_label = Entry(window, text=interface_details,  justify=RIGHT)
	instruction_text = "\nAccessing HTTP server in browser: http://%s:8000\nAccessing FTP server:\nUsername: user123\nPassword: pass123\nPort: 8021\nCommand in Linux: ftp %s 8021" % (IPAddr1,IPAddr1)




window.title('Python Quick Network Tool - DJENGINEER')
window.geometry("550x600")

topbar_widget = Label(window, text="Python Quick Network Tool ©DJENGINEER 2022 - " + version_number)
topbar_widget.pack(fill='both', expand=False,anchor="w")



instruction_text_widget = Text(window, height=1, width=100)
instruction_text_widget.tag_configure('tag-center', justify='center')
instruction_text_widget.tag_configure('tag-right', justify='right')
instruction_text_widget.tag_configure('tag-left', justify='left')
instruction_text_widget.pack(fill='both', expand=True,padx=10, pady=10,anchor="w")
instruction_text_widget.insert(END, interface_details,"tag-right")
instruction_text_widget.insert(END, instruction_text,"tag-left")

global status_text
status_text = "HTTP / HTTPS Server: %s\nFTP / SFTP Server: %s" % (http_status,ftp_status)
status_text_widget = Text(window, height=3, width=100)
status_text_widget.pack(fill='both',expand=False,padx=20, pady=2,anchor="w")
status_text_widget.insert(END, status_text,"tag-left")



def update_status():
	global status_text_widget
	status_text_widget.delete('1.0', END)
	status_text = "HTTP / HTTPS Server: %s\nFTP / SFTP Server: %s" % (http_status,ftp_status)
	status_text_widget.insert(END, status_text,"tag-left")
	status_text_widget.tag_config("stopped", foreground="red")
	status_text_widget.tag_config("started", foreground="green")
	if http_status == "STOPPED":
		status_text_widget.tag_add("stopped", "1.20", "1.28")
	elif http_status == "STARTED":
		status_text_widget.tag_add("started", "1.20", "1.28")
	if ftp_status == "STOPPED":
		status_text_widget.tag_add("stopped", "2.19", "2.28")
	elif ftp_status == "STARTED":
		status_text_widget.tag_add("started", "2.19", "2.28")
	window.after(1000, update_status) # every second...

# dev note
# use partial for passing args to tkinter command https://stackoverflow.com/questions/6920302/how-to-pass-arguments-to-a-button-command-in-tkinter
# or lambda?

# tkinter positioning
# https://stackoverflow.com/questions/51631105/how-to-position-several-widgets-side-by-side-on-one-line-with-tkinter

leftframe = Frame(window)
leftframe.pack(side=LEFT)
  
rightframe = Frame(window)
rightframe.pack(side=RIGHT)

start_http = Button(leftframe, text="start HTTP Server (port 8000)",command=start_http_button)
start_https = Button(leftframe, text="start HTTPS Server (port 8000)",command=start_https_button)
stop_http = Button(leftframe, text="stop HTTP/HTTPS Server",command=stop_http_button)

start_ftp = Button(rightframe, text="start FTP Server user123:pass123 port 8021",command=start_ftp_button)
start_ftps = Button(rightframe, text="start SFTP Server user123:pass123 port 8021",command=start_ftps_button)
stop_ftp = Button(rightframe, text="stop FTP/SFTP server",command=stop_ftp_button)


#start_http.grid(column=0, row=0) 
#start_https.grid(column=1, row=0) 
#stop_http.grid(column=2, row=0) 

#start_ftp.grid(column=0, row=1) 
#start_ftps.grid(column=1, row=1) 
#stop_ftp.grid(column=2, row=1) 


start_http.pack(padx=2, pady=2)
start_https.pack(padx=2, pady=2)
stop_http.pack(padx=2, pady=2)

start_ftp.pack(padx=2, pady=2)
start_ftps.pack(padx=2, pady=2)
stop_ftp.pack(padx=2, pady=2)




window.protocol("WM_DELETE_WINDOW", on_closing)



if __name__ == "__main__":
	if sys.platform.startswith('win'):
		# On Windows calling this function is necessary.
		multiprocessing.freeze_support()
	update_status()
	window.mainloop()